Interpolate tutorial: discuss the bases and interconversions

ev-br opened the pull request:
Following up on https://github.com/scipy/scipy/issues/22513#issuecomment-2660517663, update the interpolate user guide:

- remove statements which turn out to be more confusing than helpful
- add a discussion of BPoly objects
- add a minimal discussion of conversions between different bases of piecewise polynomials. @mdhaber you wanted to expand this part?

ev-br committed:
DOC: interpolate/tutorial: rm a confusing statement

The discussion of b-spline properties seems to cause more confusion
than clarity, so we either expand this section (by *a lot*) or just
defer the discussion to the literature.

ev-br committed:
DOC: interpolate/tutorial: discuss BPoly, conversions

[docs only]

j-bowhay commented on doc/source/tutorial/interpolate/splines_and_polynomials.rst:
```diff
@@ -203,10 +203,11 @@ B-splines are piecewise polynomials, represented as linear combinations of
+The b-spline basis is used in a variety of applications which include interpolation,
+regression and curve representation. The main feature is that each basis element is
+*localized* and equal to zero outside of a certain interval (see below for an illustration).
+This is in contrast to usual monomials, :math:`x^m`, which are non-zero on the whole
```
I get that this is trying to avoid introducing readers to what it means for a function to have compact support but this is maybe slightly confusing as $x^m$ is clearly zero at one point!

j-bowhay replied:
It is also slightly confusing as we are told
> *b-spline basis elements* --- which themselves are certain linear combinations
of usual monomials
But then we're told b-spline basis elements are localised but monomials aren't which if you were reading for the first time could sound a bit contradictory

ev-br replied:
Yeah, I can see how this is confusing. Any suggestions on how to make it less confusing?

ilayn replied:
I don't think we need math-speak for compact support. _"These bases can take nonzero values only on a finite interval and do not extend to infinity like polynomial basis functions do. This makes them attractive for local approximations etc."_ or something along those lines maybe mentioning wavelets or whatever.

mdhaber replied:
I think there are two distinct points of confusion above (which I actually understood, so they didn't both me) and two others that I *didn't* understand before. I will keep it short, but just say that there is an important distinction between a basis that is "the usual monomials $x^i$", $i \in {0, 1, \dots, n}$ and a basis of "piecewise (power basis) polynomials", each *segment* $i$ of which _can be thought of_ as the sum of $k + 1$ functions $f_{i, j} = x^j$, $j \in {0, 1, \dots, k}$. I think this is subtle, and easily missed unless the point is made very explicitly. If we want to avoid this section exploding, is there any way to avoid mentioning the fact that B-spline basis functions have compact support (regardless of whether we use "localized", "compact support", or something else)? Any readers like me would not appreciate that fact (without a lot of outside study), anyway.

ev-br replied:
Okay, let's shorten the discussion a bit more, and refer the reader to the literature.

ev-br commented:
Close/reopen to flush CircleCI. The doc build errored out, and I wonder if it was a fluke, or it's Sphinx 8.2.0 released on Feb 18.

ev-br committed:
MAINT/DOC: upper cap Sphinx

While at it, add `linkify-it-py` to pyproject.toml & doc/requirements.txt
Whatever it is, it seems needed for the doc build.

[docs only]

ev-br commented:
The doc build is green now after capping Sphinx to < 8.2.0 and adding a missing `linkify` requirement to the doc build environment.
The benchmark error is unrelated.

mdhaber commented:
A few final suggestions in ev-br/scipy#18.

mdhaber committed:
DOC: interpolate: tutorial adjustments

* DOC: interpolate: tutorial adjustments

Co-authored-by: Matt Haberland <mhaberla@calpoly.edu>

j-bowhay commented on doc/source/tutorial/interpolate/splines_and_polynomials.rst:
```diff
@@ -199,14 +199,17 @@ of :math:`1, x, \cdots, x^k`.
+The properties of b-splines are well described in the literature (see, for examples,
+references listed in the `BSpline` docstring). For our purposes, it is enough to know
+that a b-spline function is uniquely defined by an array of coefficients and
+and array of the so-called *knots*, which may or may not coincide with the data points,
```
```suggestion
an array of the so-called *knots*, which may or may not coincide with the data points,
```

j-bowhay commented on doc/source/tutorial/interpolate/splines_and_polynomials.rst:
```diff
@@ -286,3 +289,49 @@ contains basis elements evaluated at ``xnew[j]``:
+
+.. math::
+
+    b(t; k, a) = C_k^a t^k (1-t)^{k - a}
```
```suggestion
    b(t; k, a) = C_k^a t^a (1-t)^{k - a}
```
Possibly typo?

j-bowhay commented on doc/source/tutorial/interpolate/splines_and_polynomials.rst:
```diff
@@ -286,3 +289,49 @@ contains basis elements evaluated at ``xnew[j]``:
+basis, and b-splines) are equivalent, and a polynomial in one basis can be converted
+into a different basis. One reason for converting between bases is that not all bases
+implement all operations. For instance, root-finding is only implemented for `PPoly`,
+and therefore to find roots of a `BSpline` object, you need convert to `PPoly` first.
```
```suggestion
and therefore to find roots of a `BSpline` object, you need to convert to `PPoly` first.
```

j-bowhay commented on doc/source/tutorial/interpolate/splines_and_polynomials.rst:
```diff
@@ -199,14 +199,17 @@ of :math:`1, x, \cdots, x^k`.
-and is useful for a variety of applications which include interpolation, regression
-and curve representation. The main feature is that these basis elements are
-*localized* and equal to zero outside of an interval defined by the *knot array*.
+The properties of b-splines are well described in the literature (see, for examples,
```
```suggestion
The properties of b-splines are well described in the literature (see, for example,
```

ev-br committed:
DOC: interpolate/tutorial: fix typos

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

ev-br commented:
Thanks Jake, took your edits along.

j-bowhay reviewed (approved):
LGTM!

rgommers commented:
> The doc build is green now after capping Sphinx to < 8.2.0 and adding a missing `linkify` requirement to the doc build environment.

What's going on with `linkify-it-py` there? We don't use it, so we shouldn't be adding a dependency without understanding why. If it's needed, it has to be a bug in one of our other doc dependencies.

ev-br commented:
> What's going on with linkify-it-py there? 

I was wondering about this, too. At least locally, the doc build fails without it.

rgommers reviewed (dismissed):
> I was wondering about this, too. At least locally, the doc build fails without it.

Let's investigate then - this PR shouldn't be merged with this dependency added, unless (a) we understand why, and (b) it's temporary and we know how/when to revert it.

ev-br commented:
The extension is present in https://github.com/scipy/scipy/blob/main/doc/source/conf.py#L411, and was added in https://github.com/scipy/scipy/blob/main/doc/source/conf.py#L411

In that PR it was manually added to `doc/requirements.txt` and `environment.yml`, and here in this PR I did as directed by the top comment in the former file: https://github.com/scipy/scipy/pull/22541/files#diff-9ba32e29b4b0a2ef5581fa10c8a8a79f01605f541f722957c098fc05a4addb22R2

@melissawm could you advise?

melissawm commented:
Hi all - linkify-it-py helps create hyperlinks automatically when writing syntax like "https://example.com" in markdown files. This was added as part of https://github.com/scipy/scipy/pull/22385 to match what rst already does. If you feel this is not appropriate, feel free to remove. It will not break existing docs builds, it will only disable those links.

ev-br commented:
Okay, so the addition to pyproject.toml is just maintenance within the status quo; if we want to remove the dependency, we can do it separately.
One other unrelated doc change is the Sphinx version pin to unblock the problem which I believe also is also on main. Am happy to submit it as a separate PR, if desired.

Other than that, the PR is ready from my side.
